### ISSUE TITLE
Support for multi-output label (multitask learning)

### ISSUE THREAD

**Issue by bagustris** [felixbur/nkululeko#86]
The current behavior of Nkululeko only supports a single target (e.g., `target = emotion`). Given that [scikit-learn supports multiouput labels](https://scikit-learn.org/stable/modules/multiclass.html), it is possible to make Nkululeko also support multitask learning.  

Imagine that we can do,

```INI
target = ['emotion', 'age', 'gender']
```

in the INI file.

For the first start, the one could change the format of the target from `target = emotion` to `target = [ 'emotion']` and do the rest later.

**Comment by felixbur** [comment]
interesting-
need to look into this

**Comment by bagustris** [comment]
This issue is ideal for **master research**. I will try to see if there is a student interested in tackling this issue as their master project.

 (*If you are a master's student interested in tackling this issue, just comment below or contact me, and we will guide you*)

**Comment by felixbur** [comment]
that would be perfect, obviously.

My master students use nkululeko, but usually they can't program

**Comment by bagustris** [comment]
Review from copilot regarding this issue: 

What makes it a good fit

 - The codebase already has PyTorch-based ANN models (MLP, CNN) with a clean base class in model.py. An LSTM would follow the exact same pattern — inherit Model, call super().set_model_type("ann"), use the existing training/epoch loop in modelrunner.py.
 - The layers = [256, 128] config style proposed in the issue is already standard in MLP/CNN.
 - Only two files need touching: add model_lstm.py + one elif in modelrunner.py (or register via plugin registry once #33 is done).

The key challenge to be aware of

The current data pipeline outputs one aggregated feature vector per utterance (flat DataFrame row). LSTM expects sequential input (frames over time). There are two ways to handle this:

| Approach  | Feasibility  | Quality |
| ---------------- | --------------  | --------- |
| Treat each feature dim as a timestep (reshape flat vector → sequence)    | Easy, fits current   pipeline  | Limited — loses true temporal structure  |
| Frame-level features (e.g. from melspec, stft) reshaped to (batch, time, | Needs a data loader   | Proper LSTM usage   |


The MLP already partially does approach 1 (flat vector in). For proper speech LSTM, approach 2 is better — frame-level spectral features (mel spectrogram already exists as a feature extractor) reshaped into sequences.

Recommendation

Yes, add it — but implement it properly with frame-level input using existing melspec/stft features. The effort is modest since:

 1. The PyTorch training loop infrastructure is already there
 2. melspec feature extractor already exists and produces the right raw material
 3. It fills a real gap — MLP/CNN don't model temporal dynamics

**Comment by bagustris** [comment]
## Implementation Proposal

Adding detail to this feature request with a concrete implementation plan.

### Current Behavior

`target` in `[DATA]` accepts only a single string:

```ini
[DATA]
target = emotion
```

This value propagates as a plain string through `glob_conf.target`, `dataset.py`, `datasplitter.py`, all models, and the reporter. The entire pipeline assumes exactly one prediction target.

---

### Implementation Plan

#### Phase 1 — Normalize the config value (non-breaking, prerequisite)

Change config parsing so `target` is always treated as a list internally, even when a single string is given. This is the safe first step that makes all subsequent work possible.

**`nkululeko/experiment.py` (and `glob_conf.py`)**:
```python
import ast
raw = glob_conf.config["DATA"]["target"]
try:
    targets = ast.literal_eval(raw)
    if isinstance(targets, str):
        targets = [targets]
except (ValueError, SyntaxError):
    targets = [raw.strip()]
glob_conf.targets = targets          # new list-typed global
glob_conf.target  = targets[0]       # keep backward-compat single-target alias
```

No behavior change for existing INI files. All existing code continues to work via `glob_conf.target`.

---

#### Phase 2 — Data loading: load all target columns

**`nkululeko/data/dataset.py`**:
- `self.col_label` becomes `self.col_labels` (list)
- `db.get()` call passes all target column names
- `df["class_label"]` backup is created for each target: `df[f"class_label_{t}"]`
- Type detection (classification vs regression) runs per target

**`nkululeko/data/datasplitter.py`**:
- `label_encoder` becomes a dict: `self.label_encoders = {t: LabelEncoder() for t in targets}`
- Each target column is encoded independently

---

#### Phase 3 — Model support

**Scikit-learn models** (SVM, XGB, etc.) — wrap with `MultiOutputClassifier` / `MultiOutputRegressor` when `len(targets) > 1`:
```python
from sklearn.multioutput import MultiOutputClassifier
if len(glob_conf.targets) > 1:
    self.clf = MultiOutputClassifier(base_clf)
```

**PyTorch models** (MLP, CNN) — add multiple output heads, one per target:
```python
self.heads = nn.ModuleDict({
    t: nn.Linear(hidden_size, num_classes[t]) for t in targets
})
```
Each head uses its own loss; total loss is sum (or weighted via `[MODEL] target_weights = [1.0, 0.5, 0.5]`).

---

#### Phase 4 — Reporting and metrics

**`nkululeko/reporting/reporter.py`**:
- Compute metrics (UAR/MSE/etc.) per target
- Report results as a table: one row per target
- Existing single-target report format is preserved when only one target is given

---

### Affected Files Summary

| File | Change |
|---|---|
| `glob_conf.py` | Add `targets` (list), keep `target` (str) as alias |
| `experiment.py` | Parse target as list |
| `data/dataset.py` | Load all target columns |
| `data/datasplitter.py` | Per-target LabelEncoder |
| `modelrunner.py` | Multi-output model wrapping |
| `models/model_mlp.py` | Multi-head output |
| `reporting/reporter.py` | Per-target metrics |
| `docs/ini_file.md` | Document list syntax for `target` |

### Backward Compatibility

All existing INI configs with `target = emotion` (single string) remain fully supported — Phase 1 normalizes them to a single-element list internally.
